internal/controller/object/object.go

```diff
@@ -340,3 +354,160 @@ func setObserved(obj *v1alpha1.Object, observed *unstructured.Unstructured) erro
+// resolve some reference, e.g.: due to reference not ready, it will then return
+// error and requeue to wait for resolving it next time.
+func (c *external) resolveReferencies(ctx context.Context, obj *v1alpha1.Object) error {
+	c.logger.Debug("Resolving referencies.")
+
+	// Loop through references to resolve each referenced resource
+	for _, ref := range obj.Spec.References {
+		var kubeClient client.Client
+
+		if ref.FromObject.APIVersion == obj.APIVersion && ref.FromObject.Kind == obj.Kind {
+			// Object resource handled by provider on local cluster
+			kubeClient = c.localClient
+		} else {
+			// Other resource usually handled on remote cluster
+			// But could also be handled on local cluster
+			kubeClient = c.client
+		}
+
+		res := &unstructured.Unstructured{}
+		res.SetAPIVersion(ref.FromObject.APIVersion)
+		res.SetKind(ref.FromObject.Kind)
+		// Try to get referenced resource
+		err := kubeClient.Get(ctx, client.ObjectKey{
+			Namespace: ref.FromObject.Namespace,
+			Name:      ref.FromObject.Name,
+		}, res)
+
+		if err != nil {
+			return errors.Wrap(err, errGetReferencedResource)
+		}
+
+		// Patch fields if any
+		if err := ref.ApplyFromFieldPathPatch(res, obj); err != nil {
+			return errors.Wrap(err, errPatchFromReferencedResource)
+		}
+
+		// Add finalizer if not exists
+		finalizer := "object/" + string(obj.UID)
+		if !meta.FinalizerExists(res, finalizer) {
+			meta.AddFinalizer(res, finalizer)
+			if err := kubeClient.Update(ctx, res); err != nil {
+				return errors.Wrap(err, errFailedToAddFinalizer)
+			}
+		}
+	}
+
+	return nil
+}
+
+// removeReferenceFinalizers removes finalizers previously added to references.
+func (c *external) removeReferenceFinalizers(ctx context.Context, obj *v1alpha1.Object) {
+	c.logger.Debug("Removing finalizers from referencies.")
+
+	// Loop through references to resolve each referenced resource
+	for _, ref := range obj.Spec.References {
+		var kubeClient client.Client
+
+		if ref.FromObject.APIVersion == obj.APIVersion && ref.FromObject.Kind == obj.Kind {
+			// Object resource handled by provider on local cluster
+			kubeClient = c.localClient
+		} else {
+			// Other resource usually handled on remote cluster
+			// But could also be handled on local cluster
+			kubeClient = c.client
+		}
+
+		res := &unstructured.Unstructured{}
+		res.SetAPIVersion(ref.FromObject.APIVersion)
+		res.SetKind(ref.FromObject.Kind)
+		// Try to get referenced resource
+		err := kubeClient.Get(ctx, client.ObjectKey{
+			Namespace: ref.FromObject.Namespace,
+			Name:      ref.FromObject.Name,
+		}, res)
+
+		if err != nil {
+			c.logger.Debug(errGetReferencedResource, "error", err)
+			continue
+		}
+
+		// Remove finalizer if exists
+		finalizer := "object/" + string(obj.UID)
+		if meta.FinalizerExists(res, finalizer) {
+			meta.RemoveFinalizer(res, finalizer)
+			if err := kubeClient.Update(ctx, res); err != nil {
+				c.logger.Debug(errFailedToRemoveFinalizer, "error", err)
```

Comment:
I don't think it is a good idea to ignore error and continue, we should return it. This could cause unexpected issues which would require manual intervention. But returning error would trigger a retry and it could fix issue in the next tries, e.g. in case of a transient connectivity problem.

I would already suggest to call `removeReferenceFinalizers()` in `Delete` method instead of `isNotFound()` which would give use the possibility to return an error.

Reply:
> I don't think it is a good idea to ignore error and continue, we should return it.

I agree.

> I would already suggest to call removeReferenceFinalizers() in Delete method instead of isNotFound()

The reason I did not put in `Delete` is because, I found the reconciler will re-queue after `Delete`, then trigger another round of `Observe` against the same Object. If we remove finalizer added to referenced resource in `Delete`, it will lead to referenced resource deleted right after that. Next time when `Observe` gets called, it will always fail at reference resolving since referenced resource has gone. So, it's too early to do so in `Delete` and we should defer it as much as possible.

Actually, I found a much better place to do such thing is where after the Object finalizer is removed which is the last good time we can hook our logic. With that, the Object will be successfully deleted, so we can safely remove the finalizers added by this Object to its dependencies.

The code has been updated to reflect this change. I used a custom `Finalizer` to replace the default implementation. Both `AddFinalizer` and `RemoveFinalizer` are covered. Now the code is much better readable.